services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/attachment/common/AttachmentPersister.kt

```diff
@@ -0,0 +1,53 @@
+package fi.hel.haitaton.hanke.attachment.common
+
+import fi.hel.haitaton.hanke.ALLOWED_ATTACHMENT_COUNT
+import fi.hel.haitaton.hanke.HankeNotFoundException
+import fi.hel.haitaton.hanke.HankeRepository
+import fi.hel.haitaton.hanke.currentUserId
+import java.time.OffsetDateTime
+import mu.KotlinLogging
+import org.springframework.stereotype.Component
+import org.springframework.transaction.annotation.Transactional
+
+private val logger = KotlinLogging.logger {}
+
+/** Needed for attachment services transaction management. See usages for reasoning. */
+@Component
+class AttachmentPersister(
```

Comment:
The responsibilities of AttachmentPersister and HankeAttachmentService are not very clearly separated. Both access the database using the same repositories. There's also very similar code in both of them regarding finding a hanke.

Did you consider adding a new service between HankeAttachmentController and HankeAttachmentService instead of adding a service under HankeAttachmentService?

So something like a new HankeAttachmentUploadService, which has the addAttachment function. It could delegate to HankeAttachmentService for database access. It could do the calls to scanClient and the content service. It could also handle the functionality that's been added to the controller, i.e. finding the hanke ID and validating the attachment name and type. The functionality in the persistor could be moved to HankeAttachmentService.

Reply:
I just wanted to have a simple implementation that handles the transactions without http calls. 

I guess it would be possible to handle the transaction programmatically, but it would not be very spring boot style.

I think the separation of responsibilities in  HankeAttachmentUploadService and HankeAttachmentService would not be very intuitive / clear either. 

In short, this is a bit weird looking extra class but it is needed for transactional proxying to work. 

This being said, I will try to see if a new service between the controller and service would work.

Reply:
Fixed as suggested (as I understood it). How about now?

Reply:
I think it's cleaner now.